### ISSUE TITLE
Rightclick menu in the systray does not open the systray menu making it hard to change important settings

### ISSUE THREAD

**Issue by PerArneng** [docker/for-win#1011]
### Expected behavior

A menu should pop up when i right click the whale in the systray

### Actual behavior

No menu pops up except for an "About Docker" menu item

### Information

*AppData\Local\Docker\log.txt*
```log
[11:06:23.822][Systray        ][Error  ] Failed opening Docker menu: Object reference not set to an instance of an object.
```

*docker version*
```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:30:30 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:51:55 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

*Windows*
```
OS Name             : Microsoft Windows 10 Pro
OS Version          : 10.0.15063 N/A Build 15063
OS Manufacturer     : Microsoft Corporation
OS Configuration    : Standalone Workstation
OS Build Type       : Multiprocessor Free
System Boot Time    : 2017-08-16, 18:19:23
System Manufacturer : System manufacturer
System Model        : System Product Name
System Type         : x64-based PC
System Directory    : C:\WINDOWS\system32
System Locale       : sv;Swedish
Hotfix(s)           : 3 Hotfix(s) Installed.,[01]: KB4022405,[02]: KB4034662,[03]: KB4034674
```

### Steps to reproduce the behavior

  1. Install Docker
  2. Make sure it works in powershell with `docker run hello-world`
  3. Right click on the the whale in the systray
  4. Observe as only the about docker menu appears

**Comment by PerArneng** [comment]
Is it possible to find where i can do the changes in a file or registry instead? I really need to increase the ram of the VM . I tried to do it in Hyper-V directly but docker reset it to the docker settings (which is good)

**Comment by PerArneng** [comment]
Is it possible to change settings outside of the Ui.  I need to increase the ram of the vm.  It doesn't work if I do it in the hyper - v manager directly.

**Comment by carcarbe** [comment]
I had the same problem, I´ve fixed it uninstalling docker, and then manually deleting the Docker folders in
C:\Users\youruser\AppData\Local
C:\Users\youruser\AppData\Roaming
After that, I installed docker for windows again and the systray menu was displayed correctly.

**Comment by PerArneng** [comment]
Thanks @carcarbe 🥇 that fix worked perfectly 💯 . Still a bug though but very nice to be able to actually use docker on windows now.

**Comment by wnm3** [comment]
The work around above (uninstall / delete Docker directories / reinstall) did not work for me. Right clicking on the icon in the system tray shows the typical menu (run as administrator, properties, ...) but does not switch to a menu providing choices like Settings. Is there some program I can run from a command line to get the settings dialog to load?
![image](https://user-images.githubusercontent.com/1359024/39198175-66eedbc0-47b4-11e8-9eb2-8971832f7bc7.png)

**Comment by hoegge** [comment]
Same issue on:
Version 18.06.1-ce-win73 (19507)
Channel: stable
b0f14f1

**Comment by olihofmann** [comment]
I have the same issue on  Windows 10 and this Docker Version
`
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89
 Built:             Thu Jul 25 21:17:08 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89
  Built:            Thu Jul 25 21:17:52 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683`

**Comment by syska** [comment]
Same issue here as @olihofmann ...

**Comment by PietroPasotti** [comment]
Same issue on 
# docker system info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.2
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd gelf json-file local logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.2999.amd64fre.rs1_release_inmarket.190520-1518)
Operating System: Windows Server 2016 Datacenter Version 1607 (OS Build 14393.2999)
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 8GiB
Name: Dev-NL-VM-Piet
ID: FM4F:PHAT:MOQ3:YTUN:HT62:JTE7:4T3A:A5DF:TIEW:MANS:MJCJ:U5JD
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

**Comment by MartyTremblay** [comment]
I was hoping that https://github.com/docker/for-win/issues/1011#issuecomment-326092759 would fix it for me. No luck...

**Comment by robinchesterman** [comment]
Same issue for me on Windows - Docker desktop has been unusable now for weeks. Right-clicking on the tray icon does nothing. Clearing AppData and reinstalling didn't help.

 Containers: 5
  Running: 2
  Paused: 0
  Stopped: 3
 Images: 30
 Server Version: 19.03.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.9.184-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 1.934GiB
 Name: docker-desktop
 ID: LDRE:MV2Z:GLCF:2YQ5:UAIS:IA42:RAIK:DXLE:QPB4:S6DI:34T5:YSNR
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 44
  Goroutines: 55
  System Time: 2019-09-11T09:12:15.8809773Z
  EventsListeners: 1
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

**Comment by robinchesterman** [comment]
Reverting to 2.0.0.3 resolves this issue, so I think it's something new.

**Comment by MartyTremblay** [comment]
For whatever it's worth, I resolved this issue by uninstalling Kubernetes. I followed these directions. https://stackoverflow.com/a/53271159/9557885

**Comment by PBjornberg** [comment]
My findings are the same as @robinchesterman. Had to revert back to 2.0.0.31259 to get the systray menu working. None of the 2.x releases (or edge releases) gives me the menu.
Using Windows 10 Pro 10.0.18362.

**Comment by markheath** [comment]
Same issue for me too with 2.1.0.3 on Windows 1903 build 18362.418
None of the workarounds worked for me

**Comment by robmosca** [comment]
Same issue also in my case with 2.1.0.3 on Windows 1903 build 18362.418
None of the workarounds worked for me
I had to go back to 2.0.0.3 and now clicking with the right button works, but not with the left button.

**Comment by timroster** [comment]
Same issue with both 2.1.0.4, 2.1.0.1 - nothing insightful in logs. Docker starts (usually) after a reboot and cli is available, but no access to settings menu from systram. Running on Windows 10 Enterprise version 1903 build 18362.418.  Reverting to 2.0.0.3 gets settings menu back with right-click

**Comment by kule1015** [comment]
Same issue:
No access to the menu, and Docker start with every boot.

*Windows 10 Pro:*
```
Version: 1909
Build: 18363.535
```

*Docker Desktop info:*
```
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea
 Built:             Wed Nov 13 07:22:37 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea
  Built:            Wed Nov 13 07:29:19 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by mat007** [comment]
Can anyone willing to help debug this try this internal Edge build: https://download-stage.docker.com/win/edge/41608/Docker%20Desktop%20Installer.exe
You may be able to upgrade from your Sable version, but maybe you will have to uninstall it, sorry about that.

Then after the menu fails to open or partially opens look into `%LOCALAPPDATA%\Docker\log.txt` and there should be an error with a stack trace to copy-paste here.

Thanks!

**Comment by jaylholmes** [comment]
> For whatever it's worth, I resolved this issue by uninstalling Kubernetes. I followed these directions. https://stackoverflow.com/a/53271159/9557885

This too worked for me.  Thank you.

**Comment by kule1015** [comment]
According to [uninstall docker](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/configure-docker-daemon) section "How to uninstall docker", and @mat007 internal Edge Build i can access the gui (awesome desgin). 

But after first reboot docker was not starting. I could run `docker run hello-word` but `docker pull <Image-Name>` failed. Also i could not stop it. After a second reboot every thing was working normal, nothing to complain about.

@mat007 do you need any logs?

**Comment by hoegge** [comment]
This is still the case with version 2.2.0.4 - it often takes seconds for the systray menu to open (now on a new Intel i9 cpu) and you doubt if you clicked it and then you hit quit by accident. Why do you have to do a lot of stuff before showing the menu?

**Comment by jneumaier** [comment]
> 
> 
> For whatever it's worth, I resolved this issue by uninstalling Kubernetes. I followed these directions. https://stackoverflow.com/a/53271159/9557885

I have the same issue with version 2.2.0.4 (43472). Once I installed Kubernetes the issue only occurs occasionally but the UI reacts fast most of the times. May this be related to the retrieval of the Kubernetes sub menu entries of the Kubenetes "Context" list before rendering the right click menu?

**Comment by vort3xxx** [comment]
I do not have kubernetes installed and I'm having this issue.  It takes over 60 seconds for the system tray menu to appear.

**Comment by JessePreiner** [comment]
+1

**Comment by mat007** [comment]
I created a special build with extra logging to try to understand what is happening.
Here is what this adds on my machine to `%LOCALAPPDATA%\Docker\log.txt` when I open the systray menu:
```
[14:17:42.461][Systray           ][Debug  ] Open menu
[14:17:42.464][App               ][Debug  ] Get config
[14:17:42.464][App               ][Debug  ] Add about
[14:17:42.475][App               ][Debug  ] Add settings
[14:17:42.476][App               ][Debug  ] Add troubleshoot
[14:17:42.476][App               ][Debug  ] Add switch to...
[14:17:42.477][App               ][Debug  ] Add links
[14:17:42.478][App               ][Debug  ] Add dashboard
[14:17:42.479][App               ][Debug  ] Add hub
[14:17:42.482][App               ][Debug  ] Load hub
[14:17:42.484][Systray           ][Debug  ] Created hub item
[14:17:42.484][App               ][Debug  ] Add kubernetes
[14:17:42.649][App               ][Debug  ] Add restart/quit
[14:17:42.692][Systray           ][Debug  ] Menu open
[14:17:42.728][Systray           ][Debug  ] Populate hub
[14:17:43.174][Systray           ][Debug  ] Populate hub done
```
If you’re experiencing delays when opening the menu, could you try it and report here pasting the same piece of the log I did?
Here is the build: https://download-stage.docker.com/win/edge/44675/Docker%20Desktop%20Installer.exe

Thank you!

**Comment by jneumaier** [comment]
@mat007 Thanks for the Debug Build. 

As observed before the Kubernetes menu entries seem to be involved in the delay. Please have a look at the delays after "Add kubernetes" menu entries.

Fresh installation:

```
[11:25:32.321][Systray           ][Debug  ] Open menu
[11:25:32.321][App               ][Debug  ] Get config
[11:25:32.321][App               ][Debug  ] Add about
[11:25:32.322][App               ][Debug  ] Add settings
[11:25:32.324][App               ][Debug  ] Add troubleshoot
[11:25:32.324][App               ][Debug  ] Add switch to...
[11:25:32.325][App               ][Debug  ] Add links
[11:25:32.327][App               ][Debug  ] Add dashboard
[11:25:32.327][App               ][Debug  ] Add hub
[11:25:32.328][App               ][Debug  ] Load hub
[11:25:32.328][Systray           ][Debug  ] Created hub item
[11:25:32.328][App               ][Debug  ] Add kubernetes
[11:25:34.717][App               ][Debug  ] Add restart/quit
[11:25:34.759][Systray           ][Debug  ] Menu open
[11:25:34.771][Systray           ][Debug  ] Populate hub
[11:25:34.771][Systray           ][Debug  ] Populate hub done
```

Kubernetes enabled and started, menu responds faster sometimes:

```
[11:35:48.383][Systray           ][Debug  ] Open menu
[11:35:48.383][App               ][Debug  ] Get config
[11:35:48.383][App               ][Debug  ] Add about
[11:35:48.385][App               ][Debug  ] Add settings
[11:35:48.387][App               ][Debug  ] Add troubleshoot
[11:35:48.387][App               ][Debug  ] Add switch to...
[11:35:48.388][App               ][Debug  ] Add links
[11:35:48.391][App               ][Debug  ] Add dashboard
[11:35:48.392][App               ][Debug  ] Add hub
[11:35:48.392][App               ][Debug  ] Load hub
[11:35:48.393][Systray           ][Debug  ] Created hub item
[11:35:48.393][App               ][Debug  ] Add kubernetes
[11:35:48.477][App               ][Debug  ] Add restart/quit
[11:35:48.521][Systray           ][Debug  ] Menu open
[11:35:48.534][Systray           ][Debug  ] Populate hub
[11:35:48.535][Systray           ][Debug  ] Populate hub done
```

Kubernetes enabled and started, menu still responds slow on other occasions:

```
[11:36:24.072][Systray           ][Debug  ] Open menu
[11:36:24.072][App               ][Debug  ] Get config
[11:36:24.072][App               ][Debug  ] Add about
[11:36:24.073][App               ][Debug  ] Add settings
[11:36:24.075][App               ][Debug  ] Add troubleshoot
[11:36:24.076][App               ][Debug  ] Add switch to...
[11:36:24.078][App               ][Debug  ] Add links
[11:36:24.080][App               ][Debug  ] Add dashboard
[11:36:24.081][App               ][Debug  ] Add hub
[11:36:24.081][App               ][Debug  ] Load hub
[11:36:24.082][Systray           ][Debug  ] Created hub item
[11:36:24.083][App               ][Debug  ] Add kubernetes
[11:36:26.437][App               ][Debug  ] Add restart/quit
[11:36:26.477][Systray           ][Debug  ] Menu open
[11:36:26.492][Systray           ][Debug  ] Populate hub
[11:36:26.492][Systray           ][Debug  ] Populate hub done
```

**Comment by mat007** [comment]
Thanks for trying @jneumaier !
If you run this from a shell, does it take a long time?
```
kubectl version --client
```
?

**Comment by mat007** [comment]
OK, I’m pretty sure it’s the culprit.
@jneumaier (or anyone else), if you don’t use Kubernetes, can you try to delete (or rename) `C:\Program Files\Docker\Docker\resources\bin\kubectl.exe` to see if the menu opens quicker?
If my assumption is correct is should be instant…

**Comment by jneumaier** [comment]
> ```
> kubectl version --client
> ```

Gives me some delay, then it is fast for some executions.

> 
> OK, I’m pretty sure it’s the culprit.
> @jneumaier (or anyone else), if you don’t use Kubernetes, can you try to delete (or rename) `C:\Program Files\Docker\Docker\resources\bin\kubectl.exe` to see if the menu opens quicker?
> If my assumption is correct it should be instant…

You are right. After renaming the file I right clicked for a minute and the menu was there instantly all the time.

**Comment by mat007** [comment]
> You are right. After renaming the file I right clicked for a minute and the menu was there instantly all the time.

Nice, thanks for testing, I’ll make a proper fix based on our findings.
Hopefully that should fix it for everyone.

**Comment by rickpmartin** [comment]
I ran into this issue myself and while the culprit for the slow menu is loading the kubernetes information, **the reason that is slow is due to Windows Defender**.

I was able to reproduce the issue by just running `minikube` in a terminal (takes ~60 seconds to complete).

Disabling real-time protection resolves the issue for both `minikube` (or `kubectl`) as well as the Docker menu.

To permanently address the issue, the appropriate exclusions should be added to Windows Defender (although Docker should still gracefully handle slow responses from `kubectl`).

I don't know exactly what the exclusions are yet, but usually an exclusion on the process name is adequate.

**Comment by jneumaier** [comment]
Fixed in Edge Release "Docker Desktop Community 2.3.1.0":
https://docs.docker.com/docker-for-windows/edge-release-notes/#docker-desktop-community-2310

> Fixed a delay when opening the systray menu. Fixes docker/for-win#1011.

Works for me. Please close.

**Comment by mat007** [comment]
The fix has now been released in Docker Desktop Stable 2.3.0.3, I’ll go ahead and close this issue.
